clusterspec/clusterspec.go

```diff
@@ -0,0 +1,133 @@
+package clusterspec
+
+import (
+	"net/url"
+	"strings"
+
+	v1alpha1core "github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	v1alpha1provider "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/certs"
+	"github.com/giantswarm/microerror"
+)
+
+// Config holds required configuration values for factory to be able to
+// construct provider independent Cluster configuration based on
+// ClusterGuestConfig.
+type Config struct {
+	Calico struct {
+		CIDR   int
+		MTU    int
+		Subnet string
+	}
+
+	Docker struct {
+		Daemon struct {
+			CIDR      string
+			ExtraArgs string
+		}
+	}
+
+	Etcd struct {
+		AltNames string
+		Port     int
+		Prefix   string
+	}
+
+	Kubernetes struct {
+		API struct {
+			AltNames       string
+			ClusterIPRange string
+			InsecurePort   int
+			SecurePort     int
+		}
+
+		Domain string
+
+		Hyperkube struct {
+			Docker struct {
+				Image string
+			}
+		}
+
+		IngressController struct {
+			BaseDomain string
+			Docker     struct {
+				Image string
+			}
+			InsecurePort int
+			SecurePort   int
+		}
+
+		Kubelet struct {
+			AltNames string
+			Labels   string
+			Port     int
+		}
+
+		NetworkSetup struct {
+			Docker struct {
+				Image string
+			}
+		}
+
+		SSH struct {
+			UserList []struct {
+				Name      string
+				PublicKey string
+			}
+		}
+	}
+
+	Provider struct {
+		Kind string
+	}
+}
+
+// Factory to construct provider independent cluster specifications
+type Factory interface {
+	// New constructs provider/v1alpha1.Cluster
+	New(clusterGuestConfig v1alpha1core.ClusterGuestConfig) (*v1alpha1provider.Cluster, error)
+}
```

Comment:
Do we need an interface here? Why not just passing a function parameter where it's needed?

Reply:
The thought process here was [roughly] following:

- I want a factory function `func foobar(*v1alpha1core.ClusterGuestConfig)  v1alpha1provider.Cluster` for `certconfig` resource so that it doesn't need to know how/where the information to `Cluster` came, but it gets it so that it can do its job.
- In order to work, the factory producing `Cluster` instances needs configuration from `Viper/ConfigMap` so I created `Config`.
- To be able to provide a single factory function which implementation needs `Config`, I created a type.
- I want to validate given configuration values for presence before this factory type is used.

Problem with this:
```go

type Factory struct {
    Config *Config
}
```

is that it allows you to construct `Factory` without configuration or with configuration that might be only partial.

If I go with this:
```go

type Factory struct {
    config *Config
}
```

Construction of `Factory` is still possible without config. When introducing interface, only way to have instance of it is to use construction function that validates given `Config`. Does this make sense?

Reply:
I understand. But why would you need that? This is common in Go that you have a way to create an invalid struct w/o calling constructor function.

Example https://golang.org/pkg/bufio/#Reader. It is allowed by the language `&bufio.Reader{}` which is useless. I think every sane Go programer checks how to initialize the struct before using it.

It's maybe not beautiful, but I'd still stick to the standard.

Reply:
You have a point there. I actually noticed the same thing with `testing.T` yesterday. I'll do some adjustments.